go/vt/vttablet/tabletmanager/vreplication/vreplicator.go

```diff
@@ -558,3 +568,304 @@ func recalculatePKColsInfoByColumnNames(uniqueKeyColumnNames []string, colInfos
+			sqltypes.StringBindVariable(tableName), sqltypes.StringBindVariable(string(action)))
+		if err != nil {
+			return err
+		}
+		// Use a new DB client to avoid interfering with open transactions
+		// in the shared client as DDL includes an implied commit.
+		// We're also NOT using a DBA connection here because we want to
+		// be sure that the commit fails if the instance is somehow in
+		// READ-ONLY mode.
+		dbClient := vr.vre.dbClientFactoryFiltered()
+		if err := dbClient.Connect(); err != nil {
+			log.Errorf("unable to connect to the database when saving secondary keys for deferred creation: %v", err)
+			return err
+		}
+		defer dbClient.Close()
+		if _, err := dbClient.ExecuteFetch(insert, 1); err != nil {
+			return err
+		}
+		if _, err := dbClient.ExecuteFetch(sqlparser.String(alterDrop), 1); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (vr *vreplicator) getTableSecondaryKeys(ctx context.Context, tableName string) ([]*sqlparser.IndexDefinition, error) {
+	req := &tabletmanagerdatapb.GetSchemaRequest{Tables: []string{tableName}}
+	schema, err := vr.mysqld.GetSchema(ctx, vr.dbClient.DBName(), req)
+	if err != nil {
+		return nil, err
+	}
+	// schema should never be nil, but check to be extra safe.
+	if schema == nil || len(schema.TableDefinitions) != 1 {
+		return nil, fmt.Errorf("unexpected number of table definitions returned from GetSchema call for table %q: %d",
+			tableName, len(schema.TableDefinitions))
+	}
+	tableSchema := schema.TableDefinitions[0].Schema
+	var secondaryKeys []*sqlparser.IndexDefinition
+	parsedDDL, err := sqlparser.ParseStrictDDL(tableSchema)
+	if err != nil {
+		return secondaryKeys, err
+	}
+	createTable, ok := parsedDDL.(*sqlparser.CreateTable)
+	// createTable or createTable.TableSpec should never be nil
+	// if it was a valid cast, but check to be extra safe.
+	if !ok || createTable == nil || createTable.GetTableSpec() == nil {
+		return nil, fmt.Errorf("could not determine CREATE TABLE statement from table schema %q", tableSchema)
+	}
+
+	for _, index := range createTable.GetTableSpec().Indexes {
+		if !index.Info.Primary {
+			secondaryKeys = append(secondaryKeys, index)
+		}
+	}
+	return secondaryKeys, err
+}
+
+func (vr *vreplicator) execPostCopyActions(ctx context.Context, tableName string) error {
+	// Use a new DB client to avoid interfering with open transactions
+	// in the shared client as DDL includes an implied commit.
+	// We're also NOT using a DBA connection here because we want to be
+	// sure that the work fails if the instance is somehow in READ-ONLY
+	// mode.
+	dbClient := vr.vre.dbClientFactoryFiltered()
+	if err := dbClient.Connect(); err != nil {
+		return fmt.Errorf("unable to connect to the database when executing post copy actions: %v", err)
+	}
+	defer dbClient.Close()
+
+	// Save our connection ID so we can use it to easily KILL any
+	// running SQL action we may perform later if needed.
+	qr, err := dbClient.ExecuteFetch("select connection_id()", 1)
```

Comment:
so as I believe, `dbClient` is a pool. A firther change is needed: you must grab a specific `Conn`, with which you will issue both `select connection_id()` as well as the `alter`. Otherwise, the two could potentially run from two different connections.

Reply:
I don't see where there's any pooling involved (which could cause various problems for vreplication as state is critical). A `binlogplayer.DBClient` is used: https://github.com/vitessio/vitess/blob/main/go/vt/binlog/binlogplayer/dbclient.go#L40-L44

That dbclient wraps a `mysql.Conn`, and `mysql.Connect` ends up getting called, which has no pool and does a Dial: https://github.com/vitessio/vitess/blob/9cb518e489c7fce5653ac52449688f4e38f94890/go/vt/dbconfigs/dbconfigs.go#L173-L184

https://github.com/vitessio/vitess/blob/main/go/mysql/client.go#L49

Am I missing something?

All that being said.... I'll think about the best way to add some assurance that it's still the right connection. I could do something similar to the I_S query I was doing before...

Reply:
Double and triple checked (also talked with @rohit-nayak-ps), no connection pooling is used. So IMO the current code is good. Any reliable verification would be back to what I started with, which is also fine IMO as the `information_schema.INFO` field has the query *received* by the server (same as the `general_log` entry) and we were using the same query (`ALTER TABLE`) sent to the server in the query to get the connection ID running the `ALTER`.

So which do you prefer — what I originally had or what we have now? 🙂

Reply:
My mistake then! I personally prefer the current solution, as it does not involve a search in `processlist`, and just used a known connection ID. At your discretion!

Reply:
I agree. Will stick with the current `connection_id()` approach. ❤️